video/out/vo_dmabuf_wayland.c

```diff
@@ -0,0 +1,462 @@
+/*
+ * Based on vo_gl.c by Reimar Doeffinger.
+ *
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdbool.h>
+#include <assert.h>
+
+#include <va/va_drmcommon.h>
+#include <libavutil/common.h>
+#include <libavutil/hwcontext_drm.h>
+
+#include "config.h"
+
+#include "mpv_talloc.h"
+#include "common/common.h"
+#include "misc/bstr.h"
+#include "common/msg.h"
+#include "common/global.h"
+#include "options/m_config.h"
+#include "vo.h"
+#include "video/mp_image.h"
+
+#include "gpu/hwdec.h"
+#include "gpu/video.h"
+
+#include "video/vaapi.h"
+#include "present_sync.h"
+#include "wayland_common.h"
+#include "generated/wayland/linux-dmabuf-unstable-v1.h"
+#include "generated/wayland/viewporter.h"
+
+#define WLBUF_POOL_NUM_ALLOCATED_INIT 30
+
+struct wlbuf_pool_entry {
+    uintptr_t key;
+    struct wl_buffer *buffer;
+    struct zwp_linux_buffer_params_v1 *params;
+    uint32_t drm_format;
+    bool scheduled;
+};
+
+typedef uintptr_t (*wlbuf_pool_key_provider)(struct mp_image *src);
+typedef bool (*wlbuf_pool_dmabuf_importer)(struct vo *vo, struct mp_image *src, struct wlbuf_pool_entry* entry);
```

Comment:
You could implement the importing as a dmabuf_interop for the hwdecs, as semantically that's what the mapper mapping would do. It would allow you to only have one impl as you would reuse the hwdec_vaapi code that normalises the drmprime descriptor.